src/services/pasteEdits.ts

```diff
@@ -104,12 +107,33 @@ function pasteEdits(
                 preferences,
                 formatContext,
             };
-            forEachChild(updatedFile, function cb(node) {
-                if (isIdentifier(node) && !originalProgram?.getTypeChecker().resolveName(node.text, node, SymbolFlags.All, /*excludeGlobals*/ false)) {
-                    // generate imports
-                    importAdder.addImportForUnresolvedIdentifier(context, node, /*useAutoImportProvider*/ true);
-                }
-                node.forEachChild(cb);
+
+            // `updatedRanges` represent the new ranges that account for the offset changes caused by pasting new text and
+            // `offset` represents by how much the starting position of `pasteLocations` needs to be changed.
+            //
+            // We iterate over each updated range to get the node that wholly encloses the updated range. For each child of that node, it is checked if the
+            // identifier lies within the updated range and if it is not resolved, we try resolving it.
+            const updatedRanges: TextRange[] = [];
+            let offset = 0;
+            pasteLocations.forEach((location, i) => {
+                const oldTextLength = location.end - location.pos;
+                const textToBePasted = actualPastedText ? actualPastedText[0] : pastedText[i];
+                const startPos = location.pos + offset;
+                const endPos = startPos + textToBePasted.length;
+                updatedRanges.push({ pos: startPos, end: endPos });
+                offset += textToBePasted.length - oldTextLength;
+            });

```

Comment:
Why is it that this need to be separate pass and not inline where we are iterating over nodes in the the updated range?

Reply:
In the updatedFile, we don't know the exact range of where the pasted text exists, so doing this pass helps us get the right location of the pasted text. Then, it becomes easy to just go over the updated ranges and check for unresolved nodes. Is that what you were asking about?

Reply:
Isnt it same as :
```ts
let offset = 0;
            pasteLocations.forEach((location, i) => {
                const oldTextLength = location.end - location.pos;
                const textToBePasted = actualPastedText ? actualPastedText[0] : pastedText[i];
                const startPos = location.pos + offset;
                const endPos = startPos + textToBePasted.length;
                const range = { pos: startPos, end: endPos } satisfies TextRange;
                offset += textToBePasted.length - oldTextLength;
                const enclosingNode = findAncestor(
                    getTokenAtPosition(context.sourceFile, range.pos),
                    ancestorNode => rangeContainsRange(ancestorNode, range),
                );
                if (!enclosingNode) return;
                forEachChild(enclosingNode, function importUnresolvedIdentifiers(node) {
                    if (
                        isIdentifier(node) && rangeContainsPosition(
                            range,
                            node.getStart(updatedFile),
                        ) && !originalProgram?.getTypeChecker().resolveName(
                            node.text,
                            node,
                            SymbolFlags.All,
                            /*excludeGlobals*/ false,
                        )
                    ) {
                        return importAdder.addImportForUnresolvedIdentifier(
                            context,
                            node,
                            /*useAutoImportProvider*/ true,
                        );
                    }
                    node.forEachChild(importUnresolvedIdentifiers);
                });
            });
```